src/validate.rs

```diff
@@ -0,0 +1,81 @@
+//! Input validation API (Multi-line editing)
+
+use crate::keymap::Invoke;
+use crate::Result;
+
+/// Input validation result
+#[non_exhaustive]
+pub enum ValidationResult {
+    /// Incomplete input
+    Incomplete,
+    /// Validation fails with an optional error message. User must fix the
+    /// input.
+    Invalid(Option<String>),
+    /// Validation succeeds with an optional message
+    Valid(Option<String>),
+}
+
+pub struct ValidationContext<'i> {
+    i: &'i mut dyn Invoke,
+}
+
+impl<'i> ValidationContext<'i> {
+    pub(crate) fn new(i: &'i mut dyn Invoke) -> Self {
```

Comment:
Not sure why we need dyn trait here? It's not an a public constructor, so probably ValidationContext can do everything by itself.